core/src/components/tabs/tabs.tsx

```diff
@@ -43,7 +43,17 @@ export class Tabs implements NavOutlet {
 
   async componentWillLoad() {
     if (!this.useRouter) {
-      this.useRouter = !!document.querySelector('ion-router') && !this.el.closest('[no-router]');
+      /**
+       * JavaScript andd StencilJS use `ion-router`, while
+       * the other frameworks use `ion-router-outlet`.
+       *
+       * If either component is present then tabs will not use
+       * a basic tab-based navigation. It will use the history
+       * stack or URL updates associated with the router.
+       */
+      this.useRouter =
```

Comment:
~ I'd order the querying in the highest expected usage (left to right). For example there is more developers using the JS framework output of Ionic Framework and using `ion-router-outlet` as a result. If we query it first, we skip the operation of querying for `ion-router` and result in better runtime performance for a majority of applications.

Reply:
https://github.com/ionic-team/ionic-framework/pull/29772/commits/f0ba2d7563d7da778843c4e6fe22210187269147